### ISSUE TITLE
Add support for z/os - System Z architecture

### ISSUE THREAD

**Issue by bigendiansmalls** [radare/radare2#3417]
Working to add Systemz architecture to r2.   Some of systemz instructions are already added in capstone and have looked to build that out fully and/or just build within r2.  

Here are some documents / binaries which define or represent the platform.   

This is the principles of operations manual - it documents the architecture, memory, CPU and all the machine-level (assembly) instructions, opcodes and formats.

https://goo.gl/AkE07w

I've many examples of system z assembly programs written on my blog:

https://www.bigendiansmalls.com

Here is an example of a binary compiled on the system (debugging symbols enabled).

https://goo.gl/caSnRt

And the source code /assembly listing for that binary.

https://github.com/bigendiansmalls/misc_z_code/blob/master/hello_world.c
https://github.com/bigendiansmalls/misc_z_code/blob/master/hello_world.txt

Initially trying to get all the opcodes / listings (ch 7-10) out of the Principles of ops manual into a usable format.

For example from page 7-43 (Branch Relative and Save) (BRAS)
![image](https://cloud.githubusercontent.com/assets/13771080/10282654/97eea8c4-6b41-11e5-80a9-16fc2daf29d9.png)

Would be useful to have a grepable text (or xml/json) file that could be manipulated then for header files in various places (capstone,  r2, etc)  such as (i made this one up, but others would work too)

**```bras,[RI-b],32,{'A7',{R1:4},{'5':4},{RI2:16}}```**
where
instruction_name,[instruction format],total_length,{'constant':length,[{'constant':len}|{field_name:len}...}

Then that file can be easily used to build other header files and such.

**Comment by bigendiansmalls** [comment]
Some notes about file formats

The compiler supports GOFF format, this may not be default for all binaries, but it is well documented.  From the manuals:
![image](https://cloud.githubusercontent.com/assets/13771080/10283849/21180da2-6b47-11e5-8d23-2a2e75c8be4e.png)

http://www-01.ibm.com/support/knowledgecenter/SSLTBW_2.1.0/com.ibm.zos.v2r1.cbcux01/goff.htm?lang=en

See appendix I of this document for a good description of GOFF
https://goo.gl/K4jCIp

Here is a link to the same hello world binary with GOFF flag on for compilation
https://goo.gl/H7qH6s

Will post another note about the S370 Object and Extended Object format when I have more (that is the NOGOFF options)

**Comment by bigendiansmalls** [comment]
Character set.
z/os->system z runs on the CP1047 EBCDIC character set.  Not ASCII.  I have tranlsation tables here:

https://github.com/bigendiansmalls/CodePages

**Comment by bigendiansmalls** [comment]
This manual actually has the same info as the Princ of Operations above, but much more condensed.  Working on parsing this into a usable format.

https://goo.gl/FIu4qO

**Comment by Maijin** [comment]
This issue may be related EBCDIC <->ASCII https://github.com/radare/radare2/issues/414

**Comment by mainframed** [comment]
Hi All, 

I worked a little with @bigendiansmalls to get this file put together. Hopefully it's helpful. Here's a link to the file: https://github.com/zedsec390/Programming/blob/master/reference_summary.txt

Here's a few random entries from the file:
Format: instruction_name,[instruction format],total_length,{'constant':length,[{'constant':len}|{field_name:len}...}
A,[RX-a],32,{{'5A':8},{R1:4},{X2:4},{B2:4},{D2:12}}
AD,[RX-a],32,{{'6A':8},{R1:4},{X2:4},{B2:4},{D2:12}}
ADB,[RXE],48,{{'ED':8},{R1:4},{X2:4},{B2:4},{D2:12},{M3:4},{RESERVED:4},{'1A':8}}
SRDT,[RXF],48,{{'ED':8},{R3:4},{X2:4},{B2:4},{D2:12},{R1:4},{RESERVED:4},{'41':8}}
SRL,[RS-a],32,{{'88':8},{R1:4},{R3:4},{B2:4},{D2:12}}
SRLG,[RSY-a],48,{{'EB':8},{R1:4},{R3:4},{B2:4},{DL2:12},{DH2:8},{'0C':8}}
SRLK,[RSY-a],48,{{'EB':8},{R1:4},{R3:4},{B2:4},{DL2:12},{DH2:8},{'DE':8}}

**Comment by radare** [comment]
Please, fill the sysz disassembler issues into the capstone repo. do you need help to advance on this support? can you provide sample binaries?

**Comment by XVilka** [comment]
@bigendiansmalls @mainframed can you please make a regression tests for those opcodes? In here https://github.com/radare/radare2-regressions/tree/master/t.asm (see the examples for the other architectures).

**Comment by zhroom** [comment]
Hey everyone - Am working on it - given the modest support in capstone for SystemZ, we're working on finishing out the support there then brining it forward (see @mainframed post above).

Yes - can do with the regression tests -that shouldn't be a problem..

**Comment by bigendiansmalls** [comment]
Hey everyone - Am working on it - given the modest support in capstone for SystemZ, we're working on finishing out the support there then brining it forward (see @mainframed post above).

Yes - can do with the regression tests -that shouldn't be a problem..

**Comment by radare** [comment]
👍

**Comment by mainframed** [comment]
In doing some research in to this I've noted the following: 

* Capstone has implemented SystemZ (aka z/Architecture). 
* Unfortunately Capstone relies on LLVM tables, which only implements less then half the instructions (about 525 out of 995 are missing)

For example, using the instruction 'M'
```
Format: instruction_name,[instruction format],total_length,{'constant':length,[{'constant':len}|{field_name:len}...}
M,[RX-a],32,{{'5C':8},{R1:4},{X2:4},{B2:4},{D2:12}}
```
![image](https://cloud.githubusercontent.com/assets/1973519/11575621/dd2a202c-99c6-11e5-8be4-d2c965a33b4c.png)

This instruction doesn't exist within the LLVM tables. Testing with Capstone confirms this:
Known valid instruction A (Add (32)):
```
A,[RX-a],32,{{'5A':8},{R1:4},{X2:4},{B2:4},{D2:12}}
```
```
./test_systemz    
****************
Platform: SystemZ
Code:0x5a 0x0f 0x1f 0xff 
Disasm:
0x1000:	a	%r0, 0xfff(%r15, %r1)
	op_count: 2
		operands[0].type: REG = 0
		operands[1].type: MEM
			operands[1].mem.base: REG = 1
			operands[1].mem.index: REG = 15
			operands[1].mem.disp: 0xfff
```
Missing Instruction M (Multiply (64 <- 32):
```
M,[RX-a],32,{{'5C':8},{R1:4},{X2:4},{B2:4},{D2:12}}
```
```
./test_systemz    
****************
Platform: SystemZ
Code:0x5c 0xa0 0x80 0x2c 
ERROR: Failed to disasm given code!

```

My understanding of LLVM is limited but I've put in an email in to the llvm-dev mailing list to see if there's something I'm missing.

**Comment by radare** [comment]
shouldnt this issue go for capstone?

**Comment by XVilka** [comment]
@bigendiansmalls btw, any news regarding this one?

**Comment by bigendiansmalls** [comment]
Yes - I'm working on this as a stand alone disassembler, vs waiting on llvm .

**Comment by Maijin** [comment]
keystone is able to assemble zOS looks like ?

**Comment by bigendiansmalls** [comment]
@Maijin cheers - I had not seen that, I will look this week -maybe it'll speed up the development :)

**Comment by Maijin** [comment]
And keystone Is already included in r2 at least:

```
a___  32         sysz.ks     GPL     SystemZ keystone assembler (S390X)
```

**Comment by Maijin** [comment]
@bigendiansmalls See also this: https://github.com/radare/radare2-bindings/blob/master/libr/lang/p/test-py-asm.py

**Comment by radare** [comment]
hello?

**Comment by bigendiansmalls** [comment]
Hi. So I've asked the capstone folks to pull in the latest llvm which has many more Z instructions - much better base.

https://github.com/aquynh/capstone/issues/563

No word yet. But that would be really helpful.

**Comment by radare** [comment]
👍

**Comment by XVilka** [comment]
Were there any updates on z/OS support?

**Comment by radare** [comment]
We dont even have binaries in r2r

**Comment by mainframed** [comment]
What do you mean by binaries? I can easily provide those.

**Comment by radare** [comment]
Some binaries that can be commited in the regressions repo (shareable) to be used for tests and improve r2

**Comment by XVilka** [comment]
Pull Request for capstone https://github.com/aquynh/capstone/pull/1230

**Comment by XVilka** [comment]
@mainframed please provide sample binaries.

**Comment by bigendiansmalls** [comment]
Tell me where to add the binary and I'll send a test one; then if that works; I'll send a few more formats.

**Comment by radare** [comment]
https://github.com/radare/radare2-regressions/tree/master/bins

**Comment by XVilka** [comment]
@bigendiansmalls @mainframed updated SystemZ support was merged in the capstone, can you please help to test it?

**Comment by radare** [comment]
can you provide a binary for testing in r2r?

**Comment by bigendiansmalls** [comment]
See PR#1553

**Comment by bigendiansmalls** [comment]
See PR radare/radare2-regressions/pull/1553

**Comment by radare** [comment]
any updates on this issue?

**Comment by XVilka** [comment]
@radare we're waiting now for https://github.com/aquynh/capstone/pull/1305 to land in capstone master.

**Comment by XVilka** [comment]
@radare seems now it is in the next branch, so we just should switch back to next instead of master and we are fine.

**Comment by radare** [comment]
I dont understand the way branches are used in capstone . So next is the new master again?

**Comment by radare** [comment]
we are using 4.1 already, whats the excuse now for not solving this after 4 years?

**Comment by bigendiansmalls** [comment]
Will have a look.

On Jan 3, 2019, at 06:28, Anton Kochkov <notifications@github.com<mailto:notifications@github.com>> wrote:

Should be solved already. @bigendiansmalls please confirm. The rest missing
support can be done as a separate issues if any.

**Comment by XVilka** [comment]
It seems IBM offers bounties for supporting SystemZ: https://www.bountysource.com/teams/ibm/bounties

We should apply and find a student for this bounty to work on it this summer.

**Comment by XVilka** [comment]
@bigendiansmalls can you please recheck the issue?

**Comment by bigendiansmalls** [comment]
Yep - I'll build the latest version and test it on a few binaries and report back

**Comment by bigendiansmalls** [comment]
No joy out of the box.  Using latest r2 - I tried against a  31-bit zos-compiled object and binary module (pre linking and post linking), and neither produced anything useful.  See the screenshot. I'll keep trying offsets and see if I can get it to dump useable disassembly.
![image](https://user-images.githubusercontent.com/13771080/66036456-29c6c380-e505-11e9-8374-8350bb6267c9.png)

**Comment by radare** [comment]
Can you share the binary? You can debug this with the R_DEBUG_ASSERT=1 envvar

**Comment by unixfreaxjp** [comment]
I am working a lot with SystemZ on SuSE Linux from IBM Z series mainframes, gdb and binutils are still the main tools I used. We don’t think capstone is fully supported to the platform yet.

**Comment by bigendiansmalls** [comment]
I can - do you have an email I can send it to? 

> Can you share the binary? You can debug this with the R_DEBUG_ASSERT=1 envvar
> […](#)
> On 2 Oct 2019, at 12:13, Bigendian Smalls ***@***.***> wrote: ﻿ No joy out of the box. Using latest r2 - I tried against a 31-bit zos-compiled object and binary module (pre linking and post linking), and neither produced anything useful. See the screenshot. I'll keep trying offsets and see if I can get it to dump useable disassembly. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by radare** [comment]
pancake@nopcode.org <mailto:pancake@nopcode.org>

the one that appears in the commits :)

**Comment by bigendiansmalls** [comment]
:)

```[0x00000000]> pd 10
WARNING: r_reg_get: assertion 'reg && name' failed (line 296)
Backtrace 10 stack frames.
0   libr_util.dylib                     0x000000010ebb9f74 r_sys_backtrace + 36
1   libr_util.dylib                     0x000000010ec04d05 r_assert_log + 405
2   libr_reg.dylib                      0x00000001105ebac0 r_reg_get + 80
3   libr_reg.dylib                      0x00000001105ebc10 r_reg_getv + 48
4   libr_core.dylib                     0x000000010f86e1f9 ds_print_esil_anal_init + 73
5   libr_core.dylib                     0x000000010f8697d9 r_core_print_disasm + 1737
6   libr_core.dylib                     0x000000010f76916d cmd_print + 16861
7   libr_core.dylib                     0x000000010f847b33 r_cmd_call + 611
8   libr_core.dylib                     0x000000010f781ea1 r_core_cmd_subst_i + 16305
9   libr_core.dylib                     0x000000010f741bbc r_core_cmd_subst + 1596
[1]    45037 illegal hardware instruction  r2 ./HWORLD```
> pancake@nopcode.org <mailto:pancake@nopcode.org> the one that appears in the commits :)

**Comment by radare** [comment]
I will have some time tonight to check this. Thanks!

**Comment by XVilka** [comment]
Related:
- https://github.com/aquynh/capstone/issues/1545
- https://github.com/aquynh/capstone/issues/1576

**Comment by unixfreaxjp** [comment]
FYI. Malicious ELF executable for Z/OS has just been spotted recently, Linux. ELF. Native compiled.

**Comment by bigendiansmalls** [comment]
ELF isn’t a recognized format on Z.

On May 26, 2020, at 16:11, unixfreaxjp <notifications@github.com> wrote:

﻿

FYI. Malicious ELF executable for Z/OS has just been spotted recently, Linux. ELF. Native compiled.

**Comment by unixfreaxjp** [comment]
> ELF isn’t a recognized format on Z.

My statement is insufficiently written, I mean:  in the z computer that runs some lpars and upper it z/vm is running on Linux (in this case SUSE) and ELF are in there. It is r IBM system z and I have to go in their test workplace in IMB Japan in Tokyo just to dissect (debug) it.

I am on this forensics case, not ranting.

**Comment by ret2libc** [comment]
This issue has been moved from radareorg/radare2 to radareorg/ideas as we are trying to clean our backlog and this issue has probably been created a long while ago. This is an effort to help contributors understand what are the actionable items they can work on, prioritize issues better and help users find active/duplicated issues more easily. If this is not an enhancement/improvement/general idea but a bug, feel free to ask for re-transfer to main repo. Thanks for your understanding and contribution with this issue.
